source/core/Native.cs

```diff
@@ -0,0 +1,711 @@
+                {
+                    unsafe
+                    {
+                        return *(float*)(value);
+                    }
+
+                }
+                if (type == typeof(double))
+                {
+                    unsafe
+                    {
+                        return (double)(*(float*)(value));
+                    }
+   
+                }
+
+                if (type == typeof(string))
+                {
+                    var address = (char*)(value);
+
+                    if (address == null)
+                    {
+                        return String.Empty;
+                    }
+
+                    return StringFromNativeUtf8(new IntPtr(address));
+                }
+
+                if (type == typeof(IntPtr))
+                {
+                    unsafe
+                    {
+                        return new IntPtr((long)(value));
+                    }
+                }
+
+                if (type == typeof(Math.Vector2))
+                {
+                    unsafe
+                    {
+                        var data = (float*)(value);
+
+                        return new Math.Vector2(data[0], data[2]);
+                    }
+                }
+                if (type == typeof(Math.Vector3))
+                {
+                    unsafe
+                    {
+                        var data = (float*)(value);
+
+                        return new Math.Vector3(data[0], data[2], data[4]);
+                    }
+                }
+
+                if (typeof(INativeValue).IsAssignableFrom(type))
+                {
+                    unsafe
+                    {
+                        // Warning: Requires classes implementing 'INativeValue' to repeat all constructor work in the setter of 'NativeValue'
+                        var result = (INativeValue)(System.Runtime.Serialization.FormatterServices.GetUninitializedObject(type));
+                        result.NativeValue = *value;
+
+                        return result;
+                    }
+
+                }
+
+                throw new InvalidCastException(String.Concat("Unable to cast native value to object of type '", type.FullName, "'"));
+            }
+
+            internal static string StringFromNativeUtf8(IntPtr nativeUtf8)
+            {
+                unsafe
+                {
+                    byte* address = (byte*)nativeUtf8.ToPointer();
+                    int len = 0;
+
+                    while (*(address + len) != 0)
+                        ++len;
+
+                    if (len == 0)
+                        return string.Empty;
+
+                    byte[] buffer = new byte[len];
+                    Marshal.Copy(nativeUtf8, buffer, 0, buffer.Length);
+                    return System.Text.Encoding.UTF8.GetString(buffer);
+                }
+            }
+        }
+        #endregion
+
+        #region Global Variables
+        public struct GlobalVariable
```

Comment:
We should consider making most methods of this call `unsafe`, since using it wrongly can cause unsafe code to run followed by memory access exceptions.

Reply:
That idea sounds nice, but I believe we shouldn't make `MemoryAddress` unsafe at least, it means VB.NET scripts can't do the all operations by themselves that C# script can.